api/errors/codes/codes.go

```diff
@@ -0,0 +1,331 @@
+	// `OutOfRange` if asked to read from an offset past the current
+	// file size.
+	//
+	// There is a fair bit of overlap between `FailedPrecondition` and
+	// `OutOfRange`.  We recommend using `OutOfRange` (the more specific
+	// error) when it applies so that callers who are iterating through
+	// a space can easily look for an `OutOfRange` error to detect when
+	// they are done.
+	//
+	// HTTP Mapping: 400 Bad Request
+	// Google API Mapping: Code.OUT_OF_RANGE
+	OutOfRange Code = 11
+
+	// Unimplemented means the operation is not implemented or is not
+	// supported/enabled in this service.
+	//
+	// HTTP Mapping: 501 Not Implemented
+	// Google API Mapping: Code.UNIMPLEMENTED
+	Unimplemented Code = 12
+
+	// Internal means an internal error. This means that some invariants expected
+	// by the underlying system have been broken. This error code is reserved
+	// for serious errors.
+	//
+	// HTTP Mapping: 500 Internal Server Error
+	// Google API Mapping: Code.INTERNAL
+	Internal Code = 13
+
+	// Unavailable means the service is currently unavailable. This is most likely a
+	// transient condition, which can be corrected by retrying with
+	// a backoff.
+	//
+	// See the guidelines above for deciding between `FailedPrecondition`,
+	// `Aborted`, and `Unavailable`.
+	//
+	// HTTP Mapping: 503 Service Unavailable
+	// Google API Mapping: Code.UNAVAILABLE
+	Unavailable Code = 14
+
+	// DataLoss means unrecoverable data loss or corruption.
+	//
+	// HTTP Mapping: 500 Internal Server Error
+	// Google API Mapping: Code.DATA_LOSS
+	DataLoss Code = 15
+
+	// Unauthenticated means the request does not have valid authentication
+	// credentials for the operation.
+	//
+	// HTTP Mapping: 401 Unauthorized
+	// Google API Mapping: Code.UNAUTHENTICATED
+	Unauthenticated Code = 16
+
+	// Application means there was an application error. This will typically be
+	// accompianed by a user-defined error name across the wire.
+	//
+	// HTTP Mapping: 500 Internal Server Error
+	// Google API Mapping: Code.UNKNOWN
+	Application Code = 17
+)
+
+var (
+	codeToString = map[Code]string{
+		None:               "none",
+		Cancelled:          "cancelled",
+		Unknown:            "unknown",
+		InvalidArgument:    "invalid-argument",
+		DeadlineExceeded:   "deadline-exceeded",
+		NotFound:           "not-found",
+		AlreadyExists:      "already-exists",
+		PermissionDenied:   "permission-denied",
+		ResourceExhausted:  "resource-exhausted",
+		FailedPrecondition: "failed-precondition",
+		Aborted:            "aborted",
+		OutOfRange:         "out-of-range",
+		Unimplemented:      "unimplemented",
+		Internal:           "internal",
+		Unavailable:        "unavailable",
+		DataLoss:           "data-loss",
+		Unauthenticated:    "unauthenticated",
+		Application:        "application",
+	}
+	stringToCode = map[string]Code{
+		"none":                None,
+		"cancelled":           Cancelled,
+		"unknown":             Unknown,
+		"invalid-argument":    InvalidArgument,
```

Comment:
I’m a fan of your approach here. I like that you’re using tests to verify them and that they won’t pay for them at init() time.